00-programming-fundamentals/debugging-with-vs-code.md

```diff
@@ -0,0 +1,177 @@
+
+![In the editor, lines 4 and seven have bright red circles in front, indicating that have breakpoints set](images/breakpoints_vs_code.png "In the editor, lines 4 and seven have bright red circles in front, indicating that have breakpoints set")
+
+## Running the Debugger
+
+With all that done, it's time to see this tool in action. Click the green play button, or use `f5` to start the debugger. Your screen may look something like this:
+
+![VS Code pane, the debugger has started. The line that has the breakpoint is highlighted in yellow, and the command tray at the top contains the buttons that allows users to manually step, continue or stop.](images/debugger_pause_1.png "VS Code pane, the debugger has started. The line that has the breakpoint is highlighted in yellow, and the command tray at the top contains the buttons that allows users to manually step, continue or stop.")
+
+There are a few things to highlight here. Firstly, the code has paused on the highlighted line (line 4). The breakpoint has stopped the code here before it finished executing.
+
+Next, the left hand side of our window has been filled with various useful information. The top pane stores all sorts of useful information about variables that are accessible in the current scope. 
+
+The watch window is useful, but for now we will ignore it.
+
+Below the watch window is the Call Stack. The call stack shows every method we had to call in order to get here (especially useful if you have a function that is called in more than one place). 
+
+Finally the lower left of the pane shows us a list of all the breakpoints we have set. If we want to skip one of them, we can click it's checkbox in order to ignore it for the time being.
+
+In the top right of our screen, we see a new drop-down tray:
+
+![A new tray of options. Fron left to right, the buttons mean 'Move This Panel', 'Continue', 'Step Over', 'Step Into Function', 'Step Out', 'Restart Program' and 'Stop Program'](images/debug_tray.png "A new tray of options. Fron left to right, the buttons mean 'Move This Panel', 'Continue', 'Step Over', 'Step Into Function', 'Step Out', 'Restart Program' and 'Stop Program'")
+
+From Left to Right, here's what they do, and a hotkey to do them without clicking:
+
+- Move this UI element by clicking-and-dragging (No Keyboard Shortcut)
+- Continue the program until the next breakpoint, AKA `Continue`. `f5` 
+- "Step" to the next line of code in this file, AKA `Step Over`. `f10`
+- "Step" into a method on the current line. If there are multiple possible functions, it will step into the one that is evaluated first. AKA `Step Into`. `f11`
+- "Step" out of the current function. Any code that hasn't yet been run in the function will finish, and you are returned to the line where the function was called. AKA `Step Out` `⇧f11`
+- Restart the debugging from the beginning, AKA `Restart`. (No Keyboard Shortcut)
+- Stop the current program without finishing. `⇧f5`
```

Comment:
Most macbooks are configured out of the box to override function keys with like.... changing brightness and volume and stuff like that, so I'm assuming that VS Code respects that. Although it's definitely a sidebar ("Hey! My function keys aren't working!), it's definitely worth considering adding a tiny section about that somewhere. Otherwise... I will just talk about it out loud